### ISSUE TITLE
unicode regex vs non-unicode regex

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#6268]
Migrated from [rt.perl.org#20667](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=20667) (status was 'resolved')

Searchable as RT20667$

**Comment by p5pRT** [comment]
### From jorda@edpsciences.net

#### Created by jorda@edpsciences.org



UNDER LINUX\, perl 5\.6 \(see config below\)
Running the following program leads to a 'Segmentation fault' :


\#\! /usr/bin/perl \-w


$toto = 'Hello';
$toto =~ /\\w/; \# this line provokes the problem\!


$name = 'A B';
if \($name =~ /\(\\p\{IsUpper\}\) \(\\p\{IsUpper\}\)/\)\{
&nbsp;    print "It's good\! : $1 $2  \\n";
\} else \{
&nbsp;    print "It's not good\.\.\.\\n";
\}
\# end of the program


UNDER SunOS\,with perl 5\.8

This leads to "It's not good\.\.\."\,
instead of the expected "It's good\! A B"\. if I suppress 
$toto =~ /\\w/; 
I get the right result\.



CONFIG Perl  SunOS&#8203;:
\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-\-

$>perl \-V
Summary of my perl5 \(revision 5\.0 version 8 subversion 0\) configuration&#8203;:
&nbsp;  Platform&#8203;:
&nbsp;    osname=solaris\, osvers=2\.6\, archname=sun4\-solaris
&nbsp;    uname='sunos aragon 5\.6 generic\_105181\-05 sun4u sparc sunw\,ultra\-250 '
&nbsp;    config\_args='\-Dcc=gcc'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
&nbsp;    useperlio=define d\_sfio=undef uselargefiles=define usesocks=undef
&nbsp;    use64bitint=undef use64bitall=undef uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='gcc'\, ccflags ='\-I/usr/local/include \-I/opt/local/include \-D\_LARGEFILE\_SOURCE \-D\_FILE\_OFFSET\_BITS=64'\,
&nbsp;    optimize='\-O'\,
&nbsp;    cppflags='\-I/usr/local/include \-I/opt/local/include'
&nbsp;    ccversion=''\, gccversion='2\.8\.1'\, gccosandvers='solaris2\.6'
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=4321
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
&nbsp;    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=8\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='gcc'\, ldflags =' \-L/usr/local/lib \-L/opt/local/lib '
&nbsp;    libpth=/usr/local/lib /opt/local/lib /usr/lib /usr/ccs/lib
&nbsp;    libs=\-lsocket \-lnsl \-ldb \-ldl \-lm \-lc
&nbsp;    perllibs=\-lsocket \-lnsl \-ldl \-lm \-lc
&nbsp;    libc=\, so=so\, useshrplib=false\, libperl=libperl\.a
&nbsp;    gnulibc\_version=''
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=so\, d\_dlsymun=undef\, ccdlflags=' '
&nbsp;    cccdlflags='\-fPIC'\, lddlflags='\-G \-L/usr/local/lib \-L/opt/local/lib'


Characteristics of this binary \(from libperl\)&#8203;: 
&nbsp;  Compile\-time options&#8203;: USE\_LARGE\_FILES
&nbsp;  Built under solaris
&nbsp;  Compiled at Jan  2 2003 10&#8203;:24&#8203;:51
&nbsp;  %ENV&#8203;:
&nbsp;    PERL5LIB="/home2/web/Perl&#8203;:/usr/local/lib/perl5/site\_perl/EDPS"
&nbsp;  @&#8203;INC&#8203;:
&nbsp;    /home2/web/Perl
&nbsp;    /usr/local/lib/perl5/site\_perl/EDPS
&nbsp;    /usr/local/lib/perl5/5\.8\.0/sun4\-solaris
&nbsp;    /usr/local/lib/perl5/5\.8\.0
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.8\.0/sun4\-solaris
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.8\.0
&nbsp;    /usr/local/lib/perl5/site\_perl

<details><summary>Perl Info</summary>

```

Flags:
    category=core
    severity=medium

Site configuration information for perl v5.6.1:

Configured by root at Fri Jun  7 11:06:35 PDT 2002.

Summary of my perl5 (revision 5.0 version 6 subversion 1) configuration:
  Platform:
    osname=linux, osvers=2.4.18, archname=i386-linux
    uname='linux midas 2.4.18 #4 fri may 31 01:25:31 pdt 2002 i686 unknown '
    config_args='-de -Dprefix=/usr -Dcccdlflags=-fPIC -Dd_dosuid -Darchname=i386-linux -Dprivlib=/usr/lib/perl5 -Darchlib=/usr/lib/perl5/i386-linux -Dsitelib=/usr/lib/perl5/site_perl -Dsitearch=/usr/lib/perl5/site_perl/i386-linux'
    hint=recommended, useposix=true, d_sigaction=define
    usethreads=undef use5005threads=undef useithreads=undef usemultiplicity=undef
    useperlio=undef d_sfio=undef uselargefiles=define usesocks=undef
    use64bitint=undef use64bitall=undef uselongdouble=undef
  Compiler:
    cc='cc', ccflags ='-fno-strict-aliasing -I/usr/local/include -D_LARGEFILE_SOURCE -D_FILE_OFFSET_BITS=64',
    optimize='-O2',
    cppflags='-fno-strict-aliasing -I/usr/local/include'
    ccversion='', gccversion='2.95.3 20010315 (release)', gccosandvers=''
    intsize=4, longsize=4, ptrsize=4, doublesize=8, byteorder=1234
    d_longlong=define, longlongsize=8, d_longdbl=define, longdblsize=12
    ivtype='long', ivsize=4, nvtype='double', nvsize=8, Off_t='off_t', lseeksize=8
    alignbytes=4, usemymalloc=n, prototype=define
  Linker and Libraries:
    ld='cc', ldflags =' -L/usr/local/lib'
    libpth=/usr/local/lib /lib /usr/lib
    libs=-lnsl -ldl -lm -lc -lcrypt -lutil
    perllibs=-lnsl -ldl -lm -lc -lcrypt -lutil
    libc=/lib/libc-2.2.5.so, so=so, useshrplib=false, libperl=libperl.a
  Dynamic Linking:
    dlsrc=dl_dlopen.xs, dlext=so, d_dlsymun=undef, ccdlflags='-rdynamic'
    cccdlflags='-fPIC', lddlflags='-shared -L/usr/local/lib'

Locally applied patches:
    


@INC for perl v5.6.1:
    /home2/jorda/Perl
    /home2/jorda/Proto1/PERL
    /usr/lib/perl5/i386-linux
    /usr/lib/perl5
    /usr/lib/perl5/site_perl/i386-linux
    /usr/lib/perl5/site_perl
    /usr/lib/perl5/site_perl
    .


Environment for perl v5.6.1:
    HOME=/home2/jorda
    LANG (unset)
    LANGUAGE (unset)
    LC_ALL=fr_FR
    LD_LIBRARY_PATH (unset)
    LOGDIR (unset)
    PATH=/home2/jorda/bin:/usr/local/TeX/TL/bin/i386-linux/:/usr/local/pgsql/bin:/home2/jorda/bin:/usr/local/TeX/TL/bin/i386-linux/:/usr/local/pgsql/bin:/usr/local/bin:/bin:/usr/bin:/usr/X11R6/bin:/usr/games:/opt/www/htdig/bin:/usr/local/bin/texmfbin:/usr/local/jdk1.2.2/bin:/usr/lib/qt-3.0.4/bin:.
    PERL5LIB=/home2/jorda/Perl:/home2/jorda/Proto1/PERL
    PERL_BADLANG (unset)
    SHELL=/bin/tcsh

```
</details>

**Comment by p5pRT** [comment]
### From @nwc10

On Mon\, Feb 03\, 2003 at 11&#8203;:15&#8203;:33AM \-0000\, Jean\-Paul Jorda wrote&#8203;:

> UNDER LINUX\, perl 5\.6 \(see config below\)
> Running the following program leads to a 'Segmentation fault' :
> 
> 
> \#\! /usr/bin/perl \-w
> 
> 
> $toto = 'Hello';
> $toto =~ /\\w/; \# this line provokes the problem\!
> 
> 
> $name = 'A B';
> if \($name =~ /\(\\p\{IsUpper\}\) \(\\p\{IsUpper\}\)/\)\{
>     print "It's good\! : $1 $2  \\n";
> \} else \{
>     print "It's not good\.\.\.\\n";
> \}
> \# end of the program
> 
> 
> UNDER SunOS\,with perl 5\.8
> 
> This leads to "It's not good\.\.\."\,
> instead of the expected "It's good\! A B"\. if I suppress 
> $toto =~ /\\w/; 
> I get the right result\.

On 5\.6\.1 on x86 Linux valgrind says&#8203;:

==15740== Invalid read of size 4
==15740==    at 0x80D3826&#8203;: S\_find\_byclass \(in /usr/bin/perl5\.6\.1\)
==15740==    by 0x80D36A7&#8203;: Perl\_re\_intuit\_start \(in /usr/bin/perl5\.6\.1\)
==15740==    by 0x809B74A&#8203;: Perl\_pp\_match \(in /usr/bin/perl5\.6\.1\)
==15740==    by 0x809915F&#8203;: Perl\_runops\_standard \(in /usr/bin/perl5\.6\.1\)
==15740==    Address 0x8 is not stack'd\, malloc'd or free'd

The regexp bug provoked by the \\w is still present at patch 18644
\(ie current today\) but the coredump is absent from 5\.8\.0 \(onwards\)
on Linux\.

Nicholas Clark

**Comment by p5pRT** [comment]
### From @jhi

I think I now corrected this one \(change \#18653\)\.  If you are willing and
able to recompile 5\.8\.0 for yourself\, here's the change&#8203;:

==== //depot/perl/regexec\.c\#301 \(text\) ====
Index&#8203;: perl/regexec\.c
<details><summary>Inline Patch</summary>

```diff
--- perl/regexec.c.~1~  Tue Feb  4 17:58:56 2003
+++ perl/regexec.c      Tue Feb  4 17:58:56 2003
@@ -856,6 +856,7 @@
            PL_regdata = prog->data;
            PL_bostr = startpos;
        }
+       S_cache_re(prog);
        s = find_byclass(prog, prog->regstclass, s, endpos, startpos, 1);
        if (!s) {
 #ifdef DEBUGGING

I'm marking the problem ticket as resolved.
```
</details>

**Comment by p5pRT** [comment]
@jhi - Status changed from 'new' to 'resolved'
